### PR TITLE
Sumanth PodoCount Plugin - Human Analysis

### PR DESCRIPTION
**Pull request by SumanthDevarasetty**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+# RUN apt update && \
+#     apt install software-properties-common dirmngr && \
+#     wget -qO- https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc |  tee -a /etc/apt/trusted.gpg.d/cran_ubuntu_key.asc && \
+#     add-apt-repository "deb https://cloud.r-project.org/bin/linux/ubuntu $(lsb_release -cs)-cran40/" && \
+#     apt update && \
+#     apt install  -y  r-base 
+# RUN apt install -y software-properties-common libblas3 liblapack3 libtcl8.6 libtk8.6 && \
+#     add-apt-repository -y --enable-source --yes "ppa:marutter/rrutter4.0" && \
+#     add-apt-repository -y --enable-source --yes "ppa:c2d4u.team/c2d4u4.0+" && \
+#     apt install -y r-base
```
Comment (sayatmimar) [inline Dockerfile]: Please remove comment-out lines.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+
+RUN which  python && \
+    python --version
+# RUN curl -O https://bootstrap.pypa.io/get-pip.py && \
+# RUN curl -O https://bootstrap.pypa.io/pip/3.7/get-pip.py && \
+#     python get-pip.py && \
+#     rm get-pip.py
+#RUN R -e 'install.packages("Seurat",repos ="https://cran.r-project.org")'
+#RUN R -e 'library(Seurat)'
```
Comment (sayatmimar) [inline Dockerfile]: Ditto

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    #pip install --no-cache-dir torch==1.10  torchaudio==0.10 torchvision==0.11.1 -f https://download.pytorch.org/whl/cu111/torch_stable.html && \
+
+    #pip install --no-cache-dir 'git+https://github.com/facebookresearch/fvcore' && \
+
+    #git clone https://github.com/facebookresearch/detectron2 detectron2_repo && \
+    #git clone https://github.com/facebookresearch/detectron2.git && \
+    #python -m pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cu111/torch1.10/index.html && \
+    #python -m pip install -e detectron2 && \
+    # clean up \
```
Comment (sayatmimar) [inline Dockerfile]: Ditto, as we don't need them here.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+#RUN pip install --user 'git+https://github.com/facebookresearch/fvcore'
+# install detectron2
+#RUN git clone https://github.com/facebookresearch/detectron2 detectron2_repo
+# set FORCE_CUDA because during `docker build` cuda is not accessible
+#ENV FORCE_CUDA="1"
+# remove cuda compat
+# RUN apt remove --purge cuda-compat-10-0 --yes
+
+# pregenerate font cache
+RUN python -c "from matplotlib import pylab"
```
Comment (sayatmimar) [inline Dockerfile]: I don't think we meet matplotlib in a plugin right? We can remove this.

File: podo/PodoCount_code/get_glom_props.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 """
 import numpy as np
 import skimage as sk
+from skimage.morphology import binary_dilation, disk
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_glom_props.py]: Do we need to import these again? as you're using them from `sk`. Please remove if not needed.

File: podo/PodoCount_code/get_pod_feat_spaces.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 """
 import numpy as np
+from skimage.morphology import binary_dilation, disk
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_pod_feat_spaces.py]: ditto, we're not using `skimage` here.

File: podo/PodoCount_code/get_pod_props.py [hunk 7]
```diff
@@ -12,6 +12,7 @@
 from skimage.color import rgb2gray
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_pod_props.py]: we're not using this here and looks like you're importing everything from `sk` so we don't need to import `segmentation` again

File: podo/PodoCount_code/get_pod_props.py [hunk 8]
```diff
@@ -12,6 +12,7 @@
 from skimage.color import rgb2gray
 from stain_decon import stain_decon
 from get_boundary import get_boundary
+from skimage.morphology import binary_dilation, disk
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_pod_props.py]: Not used, please remove

File: podo/PodoCount_code/get_pod_props.py [hunk 9]
```diff
@@ -21,6 +22,19 @@ def thin_section_method(podocyte_count,pod_mask,areas,bbs,glom_area,T,area_mpp2,
...
+
+
+    # Convert string inputs to floats if they are indeed strings
+    d = float(d) if isinstance(d, str) else d
+    T = float(T) if isinstance(T, str) else T
+    #k = float(k) if isinstance(k, str) else k
+
+    # Now perform the operation
+    #D = (d - T + np.sqrt((d - T)**2 + (4 * k * d * T))) / (2 * k)
+'''
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_pod_props.py]: Please remove comment-out lines

File: podo/PodoCount_code/get_wsi_mask.py [hunk 10]
```diff
@@ -9,12 +9,26 @@
 import cv2
 from skimage import color
 from skimage.transform import resize
+from PIL import Image
+import matplotlib.pyplot as plt
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_wsi_mask.py]: we're not plotting anything, please remove.

File: podo/PodoCount_code/get_wsi_mask.py [hunk 11]
```diff
@@ -9,12 +9,26 @@
 import cv2
 from skimage import color
 from skimage.transform import resize
+from PIL import Image
+import matplotlib.pyplot as plt
+from skimage.morphology import binary_dilation, disk
```
Comment (sayatmimar) [inline podo/PodoCount_code/get_wsi_mask.py]: ditto, not used, please remove

File: podo/PodoCount_code/podocount_main_serv.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+        x_length = x_stop-x_start
+        y_length = y_stop-y_start
+
+        roi = np.array(WSI_file.read_region((y_start,x_start),0,(y_length,x_length)),dtype = 'uint8')
+        roi = roi[:,:,0:3]
+        rows,cols,dims = roi.shape
+        roi_name = '/roi_' + str(bb_iter+1) + '.png'
+        roi_path = wsi_roi_dir + roi_name
+        #roi_path = output_dir + roi_name
+        plt.imsave(roi_path,roi)
```
Comment (sayatmimar) [inline podo/PodoCount_code/podocount_main_serv.py]: Why are we saving this? Is it used somewhere?

File: podo/PodoCount_code/stain_decon.py [hunk 13]
```diff
@@ -10,6 +10,7 @@
 from skimage.color import rgb2gray
 from skimage.exposure import rescale_intensity
 import cv2
+from skimage.morphology import binary_dilation, disk
```
Comment (sayatmimar) [inline podo/PodoCount_code/stain_decon.py]: Not used, please remove.

File: podo/PodoCount_code/xml_to_mask.py [hunk 14]
```diff
@@ -4,6 +4,7 @@
 """
 import numpy as np
 import sys
+sys.path.append('..')
```
Comment (sayatmimar) [inline podo/PodoCount_code/xml_to_mask.py]: we don't need to add `..` in the path, please remove.

File: podo/cli/PodoCount/PodoCount.py [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import os
+import sys
+from ctk_cli import CLIArgumentParser
+
+
+sys.path.append("..")
```
Comment (sayatmimar) [inline podo/cli/PodoCount/PodoCount.py]: Not needed, please remove, also remove the empty space.

File: podo/cli/PodoCount/PodoCount.xml [hunk 16]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<executable>
+  <category>HistomicsTK</category>
+  <title>PodoCount</title>
+  <description>PodoCount</description>
+  <version>0.1.0</version>
+  <documentation-url>https://github.com/SarderLab/PodoCount</documentation-url>
+  <license>Apache 2.0</license>
+  <contributor>Briana Santo (SUNY Buffalo)</contributor>
```
Comment (sayatmimar) [inline podo/cli/PodoCount/PodoCount.xml]: Update with your name.

File: setup.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+setup(
+    name='podo',
+    use_scm_version={'local_scheme': prerelease_local_scheme},
+    description='A Python toolkit for Histopathology Image Analysis',
+    long_description=readme,
+    long_description_content_type='text/x-rst',
+    author='Kitware, Inc.',
+    author_email='developers@digitalslidearchive.net',
+    url='https://github.com/DigitalSlideArchive/HistomicsTK',
```
Comment (sayatmimar) [inline setup.py]: update athor, email and url please.

File: setup.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+        # dask packages
+        'dask[dataframe]>=1.1.0',
+        'distributed>=1.21.6',
+        # large image sources
+        #'large-image[sources]',
+        'girder-slicer-cli-web',
+        'girder-client',
+        # cli
+        'ctk-cli',
+        'Shapely==1.8.5.post1',
```
Comment (sayatmimar) [inline setup.py]: why this particular version is needed?

File: setup.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        'nimfa>=1.3.2',
+        'numpy==1.19.5',
+        'scipy>=0.19.0',
+        'Pillow==9.5.0',
+        'pandas==1.1.5',
+        'imageio>=2.3.0',
+        'shapely[vectorized]',
+        #'opencv-python-headless<4.7',
+        #'sqlalchemy',
+        'matplotlib',
```
Comment (sayatmimar) [inline setup.py]: We shouldnt need this, please let us know why we need to save roi's in the disc. Ideally we should avoid that.

**Review by sayatmimar — CHANGES_REQUESTED** [review]

File: setup.py (hunk 19 above)
Comment (SumanthDevarasetty) [reply setup.py]: Updated the code to not save roi's, It was for testing locally. Removed matplotlib as well.

File: podo/PodoCount_code/podocount_main_serv.py (hunk 12 above)
Comment (SumanthDevarasetty) [reply podo/PodoCount_code/podocount_main_serv.py]: It's not being used anywhere else. I think they were saving them to check the outputs locally. Updated the code.

File: podo/PodoCount_code/xml_to_mask.py (hunk 14 above)
Comment (sayatmimar) [inline podo/PodoCount_code/xml_to_mask.py]: this is not needed here. Please remove. We only add this in the path when we import within the repo, here we use external libraries.

File: podo/cli/PodoCount/PodoCount.py [hunk 20]
```diff
@@ -0,0 +1,15 @@
+import os
+import sys
+from ctk_cli import CLIArgumentParser
+
+def main(args):  
+    
+    cmd = "python3 ../PodoCount_code/podocount_main_serv.py  --type '{}' --input_image '{}' --glom_xml '{}' --basedir '{}' --slider {}  --section_thickness {} --girderApiUrl {} --girderToken {} \
+            ".format(args.type, args.input_image, args.glom_xml, args.basedir, args.slider, args.section_thickness, args.girderApiUrl, args.girderToken)
```
Comment (sayatmimar) [inline podo/cli/PodoCount/PodoCount.py]: I think we can use enumeration for `args.type` to make it more convenient for users to choose. See [here ](https://github.com/girder/slicer_cli_web/blob/458d616f2807b541b955a915cbb23e4575b8e8c3/example-average-color/average_color/average_color.xml#L19). Let's use this type.

File: podo/PodoCount_code/podocount_main_serv.py [hunk 21]
```diff
@@ -144,18 +152,25 @@
...
+        pil_image_glom = Image.fromarray(glom_mask_uint8)
+        print(pil_image_glom.size, "before pil image")
+        width, height = glom_mask.shape
+        glom_mask = pil_image_glom.resize((height, width), Image.Resampling.LANCZOS)
+        glom_mask = np.array(glom_mask)
+        glom_mask = glom_mask // 255
+
+        if args.type == 'Human_Analysis':
+            xml_counter, xml_contour, gcount, pcount, glom_pod_feat_vector, indv_pod_feats = get_pod_props(roi,glom_mask,slider,x_start,y_start,xml_counter,xml_contour, gcount, pcount,dist_mpp,area_mpp2, section_thickness, wsi_pod_seg_dir)
+        elif args.type == 'Mouse_Analysis':
```
Comment (sayatmimar) [inline podo/PodoCount_code/podocount_main_serv.py]: I would also recommend to use a `enum` class here. See [this](https://docs.python.org/3/library/enum.html) documentation. You can define the class on top of this file as (feel free to change the names):
```
from enum import Enum
class InputType(Enum):
    Mouse_Analysis = 'Mouse'
    Human_Analysis = 'Human'
```

And then you can check each type as:
`args.type == InputType.Mouse_Analysis.value` and `args.type == InputType.Human_Analysis.value`

**Review by sayatmimar — CHANGES_REQUESTED** [review]
Nice work, almost there.

**Review by sayatmimar — APPROVED** [review]
LGTM!

File: podo/PodoCount_code/podocount_main_serv.py [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+WSIs = [args.input_image]
+
+#WSI general info
+section_thickness = int(args.section_thickness)
+
+NAMES = ['gloms']
+base_dir_id = folder.split('/')[-2]
+_ = os.system("printf '\nUsing data from girder_client Folder: {}\n'".format(folder))
+gc = girder_client.GirderClient(apiUrl=args.girderApiUrl)
+gc.setToken(args.girderToken)
```
Comment (sayatmimar) [inline podo/PodoCount_code/podocount_main_serv.py]: Why we authenticate girder client twice? let's keep the top one and remove this.

**Review by sayatmimar — APPROVED** [review]

**Review by anindya-paul — APPROVED** [review]

**Pull request merged**